File: typescript/src/metadata.ts
import type { Metadata } from "./models.js";
import { Node, SPACE, textContent, walk } from "./tree.js";
import { safeUrl } from "./urls.js";

const ARTICLE_TYPES = new Set(
  "Article NewsArticle BlogPosting TechArticle ScholarlyArticle MedicalScholarlyArticle Report AnalysisNewsArticle OpinionNewsArticle ReviewNewsArticle BackgroundNewsArticle APIReference LiveBlogPosting".split(
    " ",
  ),
);

function string(value: unknown): string | null {
  return typeof value === "string" ? value.trim() || null : null;
}

function record(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function articleFromJson(value: unknown): Record<string, unknown> | null {
  const stack = [value];
  while (stack.length) {
    const item = stack.pop();
    if (Array.isArray(item)) {
      for (let index = item.length - 1; index >= 0; index--) {
        stack.push(item[index]);
      }
    } else if (record(item)) {
      const kinds = typeof item["@type"] === "string" ? [item["@type"]] : item["@type"];
      if (
        Array.isArray(kinds) &&
        kinds.some(
          (kind: unknown) => typeof kind === "string" && ARTICLE_TYPES.has(kind.split("/").at(-1)!),
        )
      ) {
        return item;
      }
      if (record(item["@graph"]) || Array.isArray(item["@graph"])) {
        stack.push(item["@graph"]);
      }
    }
  }
  return null;
}

function authorName(value: unknown): string | null {
  const names: string[] = [];
  const stack = [value];
  while (stack.length) {
    const item = stack.pop();
    if (Array.isArray(item)) {
      for (let index = item.length - 1; index >= 0; index--) {
        stack.push(item[index]);
      }
    } else {
      const name = string(record(item) ? item.name : item);
      if (name) {
        names.push(name);
      }
    }
  }
  return names.join(", ") || null;
}

function readJsonLd(node: Node): Record<string, unknown> | null {
  if (node.attrs.get("type")?.toLowerCase() !== "application/ld+json") {
    return null;
  }
  try {
    return articleFromJson(JSON.parse(textContent(node, false)));
  } catch {
    return null;
  }
}

export function readMetadata(root: Node, url: string | null, base: string | null): Metadata {
  const fields = new Map<string, string>();
  let article: Record<string, unknown> | null = null;
  let title: string | null = null;
  let language: string | null = null;
  let canonical: string | null = null;
  for (const node of walk(root)) {
    if (node.tag === "meta") {
      const key = (node.attrs.get("property") ?? node.attrs.get("name") ?? "").toLowerCase();
      const content = string(node.attrs.get("content"));
      if (content && !fields.has(key)) {
        fields.set(key, content);
      }
    } else if (node.tag === "title" && title === null) {
      title = string(textContent(node));
    } else if (node.tag === "html") {
      language = string(node.attrs.get("lang") || node.attrs.get("xml:lang"));
    } else if (
      node.tag === "link" &&
      (node.attrs.get("rel") ?? "").toLowerCase().split(SPACE).includes("canonical")
    ) {
      canonical = canonical || safeUrl(node.attrs.get("href") ?? "", base) || null;
    } else if (node.tag === "script" && article === null) {
      article = readJsonLd(node);
    }
  }
  return {
    title: fields.get("og:title") || title || string(article?.headline),
    author: fields.get("author") || authorName(article?.author),
    description:
      fields.get("description") || fields.get("og:description") || string(article?.description),
    language: language || string(article?.inLanguage) || fields.get("og:locale") || null,
    publishedTime:
      fields.get("article:published_time") || fields.get("date") || string(article?.datePublished),
    url,
    canonicalUrl: canonical,
  };
}

function localAuthor(node: Node): boolean {
  if (
    node.attrs.get("itemprop") === "author" ||
    (node.attrs.get("rel") ?? "").split(SPACE).includes("author")
  ) {
    return true;
  }
  return (node.attrs.get("class") ?? "")
    .toLowerCase()
    .split(SPACE)
    .some((token) => ["byline", "author", "p-author"].includes(token));
}

export function refineMetadata(metadata: Metadata, selected: Node[]): Metadata {
  let heading: string | null = null;
  let author = metadata.author;
  let published = metadata.publishedTime;
  for (const root of selected) {
    for (const node of walk(root)) {
      if (node.tag === "h1" && heading === null) {
        heading = string(textContent(node));
      }
      if (author === null && localAuthor(node)) {
        author = string(textContent(node));
      }
      if (
        published === null &&
        node.tag === "time" &&
        node.attrs.get("itemprop") !== "dateModified"
      ) {
        published = string(node.attrs.get("datetime")) || string(textContent(node));
      }
    }
  }
  return {
    ...metadata,
    title: heading || metadata.title,
    author: metadata.author || author,
    publishedTime: metadata.publishedTime || published,
  };
}

File: rust/src/metadata.rs
use crate::{
    Metadata,
    tree::{Id, Tree, has},
    urls,
};
use serde_json::Value;
use std::collections::HashMap;
const ARTICLE_TYPES: &str = "Article NewsArticle BlogPosting TechArticle ScholarlyArticle MedicalScholarlyArticle Report AnalysisNewsArticle OpinionNewsArticle ReviewNewsArticle BackgroundNewsArticle APIReference LiveBlogPosting";
fn string(v: &Value) -> Option<String> {
    v.as_str().and_then(optional)
}
fn optional(s: &str) -> Option<String> {
    let s = s.trim();
    if s.is_empty() { None } else { Some(s.into()) }
}
fn author_name(value: &Value) -> Option<String> {
    match value {
        Value::String(s) => optional(s),
        Value::Object(_) => string(&value["name"]),
        Value::Array(values) => optional(
            &values
                .iter()
                .filter_map(author_name)
                .collect::<Vec<_>>()
                .join(", "),
        ),
        _ => None,
    }
}
fn article(tree: &Tree, id: Id) -> Value {
    if !tree.nodes[id]
        .attr("type")
        .eq_ignore_ascii_case("application/ld+json")
    {
        return Value::Null;
    }
    let Ok(value) = serde_json::from_str::<Value>(&tree.text(id, false)) else {
        return Value::Null;
    };
    let mut stack = vec![&value];
    while let Some(item) = stack.pop() {
        match item {
            Value::Array(items) => stack.extend(items.iter().rev()),
            Value::Object(map) => {
                let kind = &item["@type"];
                let kinds = if let Value::Array(kinds) = kind {
                    kinds.iter().collect::<Vec<_>>()
                } else {
                    vec![kind]
                };
                if kinds.iter().any(|k| {
                    k.as_str()
                        .is_some_and(|s| has(ARTICLE_TYPES, s.rsplit('/').next().unwrap_or(s)))
                }) {
                    return item.clone();
                }
                if let Some(graph) = map.get("@graph")
                    && (graph.is_object() || graph.is_array())
                {
                    stack.push(graph);
                }
            }
            _ => {}
        }
    }
    Value::Null
}
pub fn read(tree: &Tree, source: Option<&str>, base: Option<&str>) -> Metadata {
    let mut fields = HashMap::new();
    let mut json = Value::Null;
    let mut title = None;
    let mut language = None;
    let mut canonical = None;
    for id in tree.walk(0) {
        let n = &tree.nodes[id];
        match n.tag.as_str() {
            "meta" => {
                let key = n
                    .attrs
                    .get("property")
                    .map(String::as_str)
                    .unwrap_or(n.attr("name"))
                    .to_lowercase();
                if let Some(content) = optional(n.attr("content")) {
                    fields.entry(key).or_insert(content);
                }
            }
            "title" if title.is_none() => title = optional(&tree.text(id, true)),
            "html" => language = optional(n.attr("lang")).or_else(|| optional(n.attr("xml:lang"))),
            "link" if has(&n.attr("rel").to_lowercase(), "canonical") && canonical.is_none() => {
                canonical = urls::safe(n.attr("href"), base).filter(|s| !s.is_empty())
            }
            "script" if json.is_null() => {
                json = article(tree, id);
            }
            _ => {}
        }
    }
    Metadata {
        title: fields
            .get("og:title")
            .cloned()
            .or(title)
            .or_else(|| string(&json["headline"])),
        author: fields
            .get("author")
            .cloned()
            .or_else(|| author_name(&json["author"])),
        description: fields
            .get("description")
            .or_else(|| fields.get("og:description"))
            .cloned()
            .or_else(|| string(&json["description"])),
        language: language
            .or_else(|| string(&json["inLanguage"]))
            .or_else(|| fields.get("og:locale").cloned()),
        published_time: fields
            .get("article:published_time")
            .or_else(|| fields.get("date"))
            .cloned()
            .or_else(|| string(&json["datePublished"])),
        url: source.map(str::to_owned),
        canonical_url: canonical,
    }
}
pub fn refine(tree: &Tree, selected: &[Id], metadata: &mut Metadata) {
    let mut heading = None;
    for &root in selected {
        for id in tree.walk(root) {
            let n = &tree.nodes[id];
            if n.tag == "h1" && heading.is_none() {
                heading = optional(&tree.text(id, true));
            }
            if metadata.author.is_none()
                && (n.attr("itemprop") == "author"
                    || has(n.attr("rel"), "author")
                    || n.attr("class")
                        .to_lowercase()
                        .split_whitespace()
                        .any(|t| has("byline author p-author", t)))
            {
                metadata.author = optional(&tree.text(id, true));
            }
            if metadata.published_time.is_none()
                && n.tag == "time"
                && n.attr("itemprop") != "dateModified"
            {
                metadata.published_time =
                    optional(n.attr("datetime")).or_else(|| optional(&tree.text(id, true)));
            }
        }
    }
    if heading.is_some() {
        metadata.title = heading;
    }
}
